framework/src/utils/MooseMeshUtils.C

```diff
@@ -165,4 +165,21 @@ getSubdomainID(const SubdomainName & subdomain_name, const MeshBase & mesh)
 
   return id;
 }
+
+Point
+meshCentroidCalculator(const ReplicatedMesh & mesh)
+{
+  Point origin_pt = Point(0.0, 0.0, 0.0);
+  Real vol_tmp = 0.0;
+  // Iterate through elements to calculate the center of mass of the mesh, which is used as the
+  // origin.
+  for (const auto & elem : as_range(mesh.elements_begin(), mesh.elements_end()))
+  {
+    Real elem_vol = elem->volume();
+    origin_pt += (elem->true_centroid()) * elem_vol;
+    vol_tmp += elem_vol;
+  }
+  origin_pt /= vol_tmp;
+  return origin_pt;
+}
```

Comment:
Yes, it was a class member function of `PolygonMeshGeneratorBase` in the original version for origin calculation. I was suggested to move it into `MooseMeshUtils` as it could be a general tool.

Reply:
I corrected this and generalized it as you suggested. I have not worked on `libMesh::Parallel::Communicator` much. Hopefully it's done correctly. 😅